internal/transcode/service.go

```diff
@@ -0,0 +1,293 @@
+package transcode
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/hbomb79/Thea/internal/event"
+	"github.com/hbomb79/Thea/internal/ffmpeg"
+	"github.com/hbomb79/Thea/internal/media"
+	"github.com/hbomb79/Thea/internal/workflow"
+	"github.com/hbomb79/Thea/pkg/logger"
+)
+
+var log = logger.Get("TranscodeServ")
+
+type (
+	dataStore interface {
+		SaveTranscode(*TranscodeTask) error
+		GetAllWorkflows() []*workflow.Workflow
+		GetMedia(uuid.UUID) *media.Container
+		GetTarget(uuid.UUID) *ffmpeg.Target
+	}
+
+	// transcodeService is Thea's solution to pre-transcoding of user media.
+	// It is responsible for some key aspects of Thea:
+	//   - Transcoding workflows for newly ingested media
+	//   - Manual transcode requests for ingested media
+	//   - Live-tracking and reporting of ongoing transcodes over the event bus
+	// 	 - Persistence of completed transcodes to the transcode store
+	transcodeService struct {
+		*sync.Mutex
+		taskWg          *sync.WaitGroup
+		config          *Config
+		tasks           []*TranscodeTask
+		consumedThreads int
+
+		eventBus  event.EventCoordinator
+		dataStore dataStore
+
+		queueChange chan bool
+		taskChange  chan uuid.UUID
+	}
+)
+
+// New creates a new transcodeService, injecting all required stores. Error is returned
+// in the configuration provided is not valid (e.g., ffmpeg path is wrong)
+func New(config Config, eventBus event.EventCoordinator, dataStore dataStore) (*transcodeService, error) {
```

Comment:
It's a bit odd to have a function that accepts an interface, but not have the interface exported. The caller basically has to guess as to whether the type being passed in satisfies the interface.

Makes me think this should either export the interface (preferred) or take a concrete type instead.

Remember that defining/documenting your package API with an interface is usually discouraged in Go. Interface development usually starts with the caller defining the minimal interface it needs and this interface lives in the calling package. The implementation doesn't even need to import it - it just needs to implement it.

Reply:
Yeah, I think exporting make sense here. In this case the transcode service only needs a small subset of the `thea/store.go` orchestrator, so having it accept the full concrete type here feels dirty (especially because importing it feels like I'm no longer keeping these services properly isolated).